smart-contracts-tutorial/first-contract.md

```diff
@@ -0,0 +1,244 @@
+# Creating and running the simplest Smart Contract
+
+This first contract will be a one-player game where the player tries to guess a number. If they guess right, they win, otherwise, they lose. The game can be played any number of times and there is no actual stake. For the purpose of keeping things simple, the number is hardcoded to be `37`
+
+
+## Set up the environment for the contract
+
+Clone the [cosmic-swingset repository](https://github.com/Agoric/cosmic-swingset):
+```sh
+git clone git@github.com:Agoric/cosmic-swingset.git
+```
+
+Follow the instruction to [build and setup cosmic-swingset](https://github.com/Agoric/cosmic-swingset#build-from-source) (~20 minutes)
+
+
+## Write a contract
+
+```sh
+# Make sure you're in the cosmic-swingset root directory
+echo $PWD
+# => /maybe/home/you/code-projects/cosmic-swingset
+
+# (optional) create a git branch
+# git branch -c my-contract-tutorial
+
+# Create the contract file
+touch lib/ag-solo/vats/guess37-contract.js
+```
+
+- copy/paste the following code in the file:
+```js
+function guess37Contract(terms, inviteMaker) {
+  const seat = harden({
+    guess(attempt) {
+      return attempt === 37 ? 'you win' : 'you lose';
+    },
+  });
+
+  return harden({
+    playerInvite: inviteMaker.make('player', seat),
+  });
+}
+
+export default guess37Contract.toString();
+```
+
+(This code is explained in a section below)
+
+The contract is materialized by the source code of the `start` function. JavaScript makes it possible to retrieve the source code of a function by calling [toString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString) on the function
+
+```
+TODO a documentation of the contract format
+- Explain the contract host
+    - Explain `terms` and `inviteMaker` arguments
+- Explain the [provided top-level variables](https://github.com/Agoric/ERTP/blob/8f42af53581821bb16a9f4e3d62603833354ef8b/core/contractHost.js#L53-L63)
+- Explain Jessie
+```
+
+Change the content of `lib/ag-solo/vats/vat-pixel.js` to be the following:
```

Comment:
At this stage, i'm tiptoeing on this topic because i have no visibility on what part of cosmic-swingset is at prototype phase and what part is intended to remain for the long term
I'm starting to have a good idea of what the current UX to write a contract is, but it's not clear whether the current UX is the intended long-term UX
https://github.com/Agoric/ERTP/issues/121 and interacting with the contract host from the UI are part of this discussion

Until there is a clear consensus, I choose to make a tutorial that works with the latest commit of cosmic-swingset (which will require work to keep up if things change)

One reason for this choice is to have something that works which is good for the reader. Another is to make visible what parts of the process are ugly/annoying and potential trigger a discussion on the topic

I'm not sure how to start the conversation about the long-term UX of contract authoring. How would you start?

> Or are you avoiding the step of adding the vat to Home?

i take advantage of the fact that the pixel vat is already added to home. This keeps the focus of the learning to contracts (and not cosmic-swingset internals)

Reply:
At least compared to the contractHost layer, cosmic-swingset is likely to be pretty stable. OTOH, we don't have a story yet about how people will supply code for new contracts, so shoe-horning it in in place of an existing contract is probably okay. Maybe we should put a template contract in place just for this purpose.

Not getting side tracked into how the contract becomes visible is the right choice. I'll talk to people about whether we should add a template contract so people don't have to figure out how to get that wired up.